sound/soc/sof/intel/hda-dai.c

```diff
@@ -603,6 +603,84 @@ static const struct snd_soc_dai_ops ipc4_hda_dai_ops = {
 	.prepare = hda_dai_prepare,
 };
 
+static int dspless_hda_dai_hw_params(struct snd_pcm_substream *substream,
+				     struct snd_pcm_hw_params *params,
+				     struct snd_soc_dai *dai)
+{
+	struct hdac_stream *hstream = substream->runtime->private_data;
+	struct hdac_ext_stream *hext_stream = stream_to_hdac_ext_stream(hstream);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *codec_dai = asoc_rtd_to_codec(rtd, 0);
+	struct hdac_bus *bus = hstream->bus;
+	struct hdac_ext_link *hlink;
+	int sig_bits;
+
+	hlink = snd_hdac_ext_bus_get_hlink_by_name(bus, codec_dai->component->name);
+	if (!hlink)
+		return -EINVAL;
+
+	/* set the hdac_stream in the codec dai */
+	snd_soc_dai_set_stream(codec_dai, hdac_stream(hext_stream), substream->stream);
+
+	if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_PLAYBACK)
+		snd_hdac_ext_bus_link_set_stream_id(hlink, hstream->stream_tag);
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		sig_bits = codec_dai->driver->playback.sig_bits;
+	else
+		sig_bits = codec_dai->driver->capture.sig_bits;
+
+	/* Save the format required for the codec */
+	hstream->format_val = snd_hdac_calc_stream_format(params_rate(params),
+							  params_channels(params),
+							  params_format(params),
+							  sig_bits, 0);
```

Comment:
is saving enough? Dont we need to call snd_hdac_ext_stream_setup() with the format?

Reply:
afaik when the HDA-DMA is coupled the format only need to be set in one place and it is going to propagate to the link side

Reply:
then what does this saving really do? is it needed at all?

Reply:
this is indeed a problem, because we set this format_val information in hda-pcm.c based on the host hw_params only.
````c
hda_dsp_pcm_hw_params	
     hstream->format_val = rate | bits | (params_channels(params) - 1);


hda_dsp_stream_hw_params
     	/* program stream format */
	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
				sd_offset +
				SOF_HDA_ADSP_REG_SD_FORMAT,
				0xffff, hstream->format_val);
````

this means we never take into account the codec_dai information. It's possible that things work by accident but it doesn't seem quite right.

Reply:
We need to save this for the FE side to use.
With DSP in picture the BE (Link) side and FE (host) side of the DSP can have different format configured and the conversion is done inside of the DSP, but when we don't have the DSP we need the codec side information used on the host side so the format is correct.

Reply:
Completely agree we need to keep this information, my point is that we shouldn't program the format twice, first with a value that ignores the codec settings and second with a value that takes codec settings into account.

I don't even know which of the two is used today, the order between FE and BE keeps changing so it's anyone's guess what is used.

Maybe we should make the first assignment in hda-pcm.c conditional on the presence of the DSP?

Reply:
Like in the first chunk change in the hda-pcm.c in this patch? At line 114

Reply:
ah yes, that will do @ujfalusi. thanks for the pointer. We can close this one.